### ISSUE TITLE
Collision using hash()

### ISSUE THREAD

**Issue by aehatxdotd** [spyder-ide/spyder#22139]
### What steps will reproduce the problem?

Executing:
        print(hash((3, 1, 5, 0, 4, 2),)+hash((5, 4, 3, 2, 1, 0),))
        print(hash((5, 1, 3, 2, 4, 0),)+hash((3, 4, 5, 0, 1, 2),))
gives identical totals, and other similar issues with permuted figures exist.  This renders invalid the assumption that the sum of hashes of 
two tuples is essentially unique to the pairs of tuple. Workaround is to turm the tuples into strings befor hashing.

## Versions

* Spyder version: 5.5.4  (standalone)
* Python version: 3.8.10 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.10
* Operating System: Windows-10-10.0.22631-SP0

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0          :  1.4.1 (OK)
chardet >=2.0.0               :  5.2.0 (OK)
cloudpickle >=0.5.0           :  3.0.0 (OK)
cookiecutter >=1.6.0          :  2.6.0 (OK)
diff_match_patch >=20181111   :  20230430 (OK)
intervaltree                  :  None (OK)
IPython >=8.12.2,<8.13.0      :  8.12.3 (OK)
jedi >=0.17.2,<0.20.0         :  0.19.1 (OK)
jellyfish >=0.7               :  1.0.3 (OK)
jsonschema >=3.2.0            :  4.21.1 (OK)
keyring >=17.0.0              :  25.1.0 (OK)
nbconvert >=4.0               :  7.16.3 (OK)
numpydoc >=0.6.0              :  1.7.0 (OK)
paramiko >=2.4.0              :  3.4.0 (OK)
parso >=0.7.0,<0.9.0          :  0.8.4 (OK)
pexpect >=4.4.0               :  4.9.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.9.8 (OK)
pygments >=2.0                :  2.17.2 (OK)
pylint >=3.1,<4               :  3.1.0 (OK)
pylint_venv >=3.0.2           :  3.0.3 (OK)
pyls_spyder >=0.4.0           :  0.4.0 (OK)
pylsp >=1.11.0,<1.12.0        :  1.11.0 (OK)
pylsp_black >=2.0.0,<3.0.0    :  2.0.0 (OK)
qdarkstyle >=3.2.0,<3.3.0     :  3.2.3 (OK)
qstylizer >=0.2.2             :  0.2.2 (OK)
qtawesome >=1.3.1,<1.4.0      :  1.3.1 (OK)
qtconsole >=5.5.1,<5.6.0      :  5.5.1 (OK)
qtpy >=2.1.0                  :  2.4.1 (OK)
rtree >=0.9.7                 :  1.2.0 (OK)
setuptools >=49.6.0           :  69.2.0 (OK)
sphinx >=0.6.6                :  7.1.2 (OK)
spyder_kernels >=2.5.1,<2.6.0 :  2.5.1 (OK)
textdistance >=4.2.0          :  4.6.1 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog                      :  4.0.0 (OK)
zmq >=24.0.0                  :  25.1.2 (OK)

# Optional:
cython >=0.21                 :  3.0.10 (OK)
matplotlib >=3.0.0            :  3.7.5 (OK)
numpy >=1.7                   :  1.24.4 (OK)
pandas >=1.1.1                :  2.0.3 (OK)
scipy >=0.17.0                :  1.10.1 (OK)
sympy >=0.7.3                 :  1.12 (OK)
```

**Comment by aehatxdotd** [comment]
Problem arose in software exploring the properties of subgroups of symmetrical groups, and my approach needed a hash representation of a permutation that was essentially unique and without relationships to other permutation hash.  I found instances of pairs of permutations for which the sum of the two hashes was shared, causing the software logic to fail, in a reproducible but hard-to-pinpoint manner.

**Comment by bessman** [comment]
1. This is obviously not a bug in Spyder.
2. It's probably not a bug at all. `hash` makes no guarantee that different objects have different hashes, though excessive collisions may be considered erroneous.
3. Use `hashlib` if your program cannot handle collisions.

**Comment by aehatxdotd** [comment]
Thanks, Alexander.  I take your point in principle, but it is a question of what one expects of a hash – and not just for me.   Ideally, a hash gives not only a very low expectation of collision, closely related to the chance of two random numbers being the same in a particular range of numbers, but also an expectation (harder to quantify) that combinations of numbers should be derived independently.   What I showed was that four hashes had a very simple relationship a+b=c+d, suggesting that the hash was created in a mathematically simplistic way, and by ill luck this caused a rather hard-to-nail-down bug in my software are (experimental group-theory).

The simple work around is to hash a derived string, and this is OK, just a bit less efficient in computing-heavy experimental programs.  The simplicity of this work-around (even though it is by no means perfect) suggests that a better algorithm could easily have been chosen for hash().

I’m a bit sensitive about this issue, having created a new database system (for a commercial X.500 system) way back in the 80’s that used a relatively simple empirically derived hashing algorithm that eventually caused a customer problem.  Fortunately, in those days, there were a number of good hashing algorithms that had grown up with public key cryptography (then a nascent technology), that I had avoided for the first design for efficiency reasons, but was then obliged to adopt because of the impact on the database.  Fortunately, developing technology absorbed the performance problem.

Anthony

Anthony Hodson ***@***.******@***.***>
Spring Lanes House, Holly Spring Lane,
Bracknell, Berks RG12 2JL
Tel: 01344 483347
Mob: 0771 360 7086
Web: www.xdotd.com<http://www.xdotd.com/> www.fosmw.com<http://www.fosmw.com/>

**Comment by ccordoba12** [comment]
Hey @aehatxdotd, thanks for reporting. Unfortunately, as @bessman said, this is not a Spyder bug so we can't help you with it here.

However, just as a quick advice, you could take a look at the hash functions from the `cryptography` module:

https://cryptography.io/en/latest/hazmat/primitives/cryptographic-hashes/

**Comment by aehatxdotd** [comment]
Thank you, Carlos, happy to leave it at that, but perhaps useful for people to note the attitude taken by  @bessman<https://github.com/bessman> – “don’t expect too much!!” – which I did, naively, as an ancient professional (working with real computers since 1960), expecting great things 64 years later.
Best wishes
Anthony

Anthony Hodson ***@***.******@***.***>
Spring Lanes House, Holly Spring Lane,
Bracknell, Berks RG12 2JL
Tel: 01344 483347
Mob: 0771 360 7086
Web: www.xdotd.com<http://www.xdotd.com/> www.fosmw.com<http://www.fosmw.com/>

**Comment by aehatxdotd** [comment]
Dear Alexander,

I wrote a little program to exhaustively detect collisions in the sum of two permutations using the hash value hash_value=hash(tuple(perm)) where perm is a list containing ( in the example I took for demonstration) such as [1,3,2,4], taken from the 24 such permutations in the very simple symmetric group S4.

Collisions in the sum of the hash-value gave 8 pairs that generated them.  I attach the summary.  SYM[7] is (of course) [[1, 0, 3, 2].  The first three collisions were three-way.

Not an impressive record for hash(), but substituting str for tuple does the trick adequately for me.  My loss was the time spent trying to work out why I was not having the results that I expected, and collisions as the problem seemed at first less likely than the possibility of a bug in my software.

Best wishes

Anthony

Anthony Hodson ***@***.******@***.***>
Spring Lanes House, Holly Spring Lane,
Bracknell, Berks RG12 2JL
Tel: 01344 483347
Mob: 0771 360 7086
Web: www.xdotd.com<http://www.xdotd.com/> www.fosmw.com<http://www.fosmw.com/>
